ipq40xx: add Wallystech DR40x9 support

robimarko opened the pull request:
Adds support for the Wallystech DR40x9 series boards.
They come in IPQ4019 and IPQ4029 versions.
IPQ4019/4029 only differ in that that IPQ4029 is the industrial version that is rated to higher temperatures.

Specifications are:
* CPU: Qualcomm IPQ40x9 (4x ARMv7A Cortex A7) at 716 MHz
* RAM: 512 MB
* Storage: 2MB of SPI-NOR, 128 MB of parallel NAND
* USB 3.0 TypeA port for users
* MiniPCI-E with PCI-E 2.0 link
* MiniPCI-E for LTE modems with only USB2.0 link
* 2 SIM card slots that are selected via GPIO11
* MicroSD card slot
* Ethernet: 2x GBe with 24~48V passive POE
* SFP port (Untested, I have no way to test it)
* DC Jack
* UART header
* WLAN: In-SoC 2x2 802.11b/g/n and 2x2 802.11a/n/ac
* 4x MMCX connectors for WLAN
* Reset button
* 8x LED-s

Installation instructions:
Connect to UART, pins are like this:
-> 3.3V | TX | RX | GND

Settings are 115200 8n1

Boot initramfs from TFTP:
tftpboot 0x84000000 openwrt-ipq40xx-generic-wallystech_dr40x9-initramfs-fit-uImage.itb

bootm

Then copy the sysupgrade image to the /tmp folder and execute sysupgrade -n <image_name>

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

blocktrron commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,374 @@
+	perst-gpio = <&tlmm 38 GPIO_ACTIVE_HIGH>;
+	wake-gpio = <&tlmm 40 GPIO_ACTIVE_HIGH>;
+
+	/* Free slot for use */
```
`Unpolulated slot` describes it better IMHO

robimarko replied:
Yeah, sounds better

blocktrron commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -36,7 +36,8 @@ ALLWIFIBOARDS:= \
 	mobipromo_cm520-79f \
-	qxwlan_e2600ac
+	qxwlan_e2600ac \
+  wallystech_dr40x9
```
Indentation is off here.

robimarko replied:
Agree, will fix it.

blocktrron commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -109,5 +110,6 @@ $(eval $(call generate-ipq-wifi-package,ezviz_cs-w3-wd1200g-eup,EZVIZ CS-W3-WD12
 $(eval $(call generate-ipq-wifi-package,linksys_ea8300,Linksys EA8300))
 $(eval $(call generate-ipq-wifi-package,mobipromo_cm520-79f,MobiPromo CM520-79F))
 $(eval $(call generate-ipq-wifi-package,qxwlan_e2600ac,Qxwlan E2600AC))
+$(eval $(call generate-ipq-wifi-package,wallystech_dr40x9,Wallystech DR40X9))
```
Is this already submitted to the ath10k mailing list?

robimarko replied:
No, but I gave up on submitting stuff as I have BDF-s that are now over 1-year old waiting to be merged.

blocktrron replied:
This happens very irregularly, i admit that. But please submit it so we can eventually drop the bdfs from our tree.

robimarko replied:
Ok, will submit them.

blocktrron commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,374 @@
+	compatible = "wallystech,dr40x9";
+
+	chosen {
+		bootargs-append = " ubi.mtd=ubi root=/dev/ubiblock0_1";
```
The kernel should automount the UBI volume. Or is the bootloader overwriting the bootline?

robimarko replied:
The bootloader is passing ubi.mtd=rootfs and different root partition so it needs to overwritten

blocktrron replied:
Is this a quirk the bootloader does when booting linux images or can this behavior turned off by changing the bootline passed by the bootloader?

If so, adjusting the installation instructions should be preferred over using the `bootargs-append` hack.

robimarko replied:
It dynamically generates the bootargs so that is why I used the bootargs-append.

T-X replied:
Btw., technically the "chosen ... bootargs-append" option works for us, too. It just looks a bit weird, to have root= / rootfs= listed twice, with the former, provided from u-boot, being ignored.

And we ran over some ethernet issues with the original u-boot (from QSDK?) + OpenWrt + DTS we got from Wallsys (looks slightly different then this one) combination. The probing fails and eth0/eth1 do not appear. Running a "tftpboot" or "ping" from u-boot usually helps until power-cycling. So it seems that the original u-boot (or driver / DTS combination) was missing some reset of the switch chip.

However after flashing the new u-boot provided by Wallsys via the "source ...:script" this worked too, it seems that the new u-boot we got did that reset.

But it seems you found the appropriate reset GPIO for the MDIO, your DTS looks better than the one we got from Wallsys in that regard. I guess that fixed exactly this issue for you?

robimarko replied:
You are over complicating this.
U-boot can read the partition table from SMEM, but that is usually completely incorrect as vendors simply use one of the reference board configs.
So they simply set the U-boot env to generate the correct bootargs and that gets passed to the kernel, so you can either modify the environment or use bootargs-append to correct them from DTS.

Yeah, its weird but Linux will always take the last arguments from cmdline so you overwrite whatever vendor did that way.

I did not really use QSDK at all, simply used it to generate a running DTS and thats it.
Yeah, without the correct PHY reset GPIO Linux can sometimes be unable to properly configure ethernet as QCA U-boot tends to do some weird stuff depending on the version used.

Wallystech always provides some messed up mainline and QSDK DTS combo, which is useless, and please don't rely on the U-boot scripts for flashing.
Those only work for QSDK images.

ecsv replied:
We are currently in contact with Wallystech to get their u-boot sources. They didn't yet provide any sources which matches their binaries and thus wasn't yet tested on a board. But the currently available things can be found at https://github.com/ecsv/u-boot-wallystech-dr4029 (see README.build for the most likely instruction how to get it build).

This can then also be used to provide a publicly available u-boot binary which just works (tm).

T-X replied:
> So they simply set the U-boot env to generate the correct bootargs and that gets passed to the kernel, so you can either modify the environment or use bootargs-append to correct them from DTS.

Ah! You're right, thanks to your previous findings [here](https://forum.openwrt.org/t/bootipq-and-bootm-difference-in-uboot-was-sysupgrade-from-initramfs-on-ipq4019/12547) regarding bootipq vs. bootm I now found a way which works without the "chosen ... bootargs-append". Seems the annoying "bootargs" is only hardcoded in the uboot for the "bootipq" command. Running the following instead boots fine without needing the bootargs-append and without needing to reflash the uboot:

```
(IPQ40xx) # printenv  
baudrate=115200
bootcmd=bootipq
bootdelay=2
ethact=eth0
flash_type=0
ipaddr=192.168.1.11   
machid=8010006
serverip=192.168.1.10 
stderr=serial
stdin=serial
stdout=serial

Environment size: 190/65532 bytes
(IPQ40xx) # nand device 0 && set mtdids nand0=nand0 && set mtdparts mtdparts=nand0:0x4000000@0x0(fs),${msmparts} && ubi part fs && ubi read 0x84000000 kernel
...
(IPQ40xx) # bootm
...
root@OpenWrt:/# cat /proc/cmdline 

root@OpenWrt:/# 
```

The "(IPQ40xx) # nand device 0 && ..." seems to be what is hardcoded into uboot's "bootipq".

T-X replied:
And, surprise, surprise, changing the "mtdparts=nand0:0x4000000@0x0(fs)" to 0x8000000 then also works for a 128MB NAND setting in the DTS and sysupgrade.

adrianschmutzler commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,374 @@
+#include "qcom-ipq4019.dtsi"
```
Please add a license (SPDX).

robimarko replied:
Done

adrianschmutzler commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,374 @@
+		};
+
+		wlan2g-strength {
+			label = "dr4029:green:wlan2g-strenght";
```
Typo "strength"

robimarko replied:
Done

adrianschmutzler commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,374 @@
+		};
+
+		wlan5g-strength {
+			label = "dr4029:green:wlan5g-strenght";
```
Typo "strength"

robimarko replied:
Done

adrianschmutzler commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,374 @@
+			};
+
+			partition5@e0000 {
+				label = "0:APPSBLENV"; /* uboot env*/
```
Nitpick: add space before */

robimarko replied:
Done

robimarko committed:
ipq40xx: add Wallys DR40x9 support

Adds support for the Wallys DR40x9 series boards.
They come in IPQ4019 and IPQ4029 versions.
IPQ4019/4029 only differ in that that IPQ4029 is the industrial version that is rated to higher temperatures.

Specifications are:
* CPU: Qualcomm IPQ40x9 (4x ARMv7A Cortex A7) at 716 MHz
* RAM: 512 MB
* Storage: 2MB of SPI-NOR, 128 MB of parallel NAND
* USB 3.0 TypeA port for users
* MiniPCI-E with PCI-E 2.0 link
* MiniPCI-E for LTE modems with only USB2.0 link
* 2 SIM card slots that are selected via GPIO11
* MicroSD card slot
* Ethernet: 2x GBe with 24~48V passive POE
* SFP port (Untested, I have no way to test it)
* DC Jack
* UART header
* WLAN: In-SoC 2x2 802.11b/g/n and 2x2 802.11a/n/ac
* 4x MMCX connectors for WLAN
* Reset button
* 8x LED-s

Installation instructions:
Connect to UART, pins are like this:
-> 3.3V | TX | RX | GND

Settings are 115200 8n1

Boot initramfs from TFTP:
tftpboot 0x84000000 openwrt-ipq40xx-generic-wallys_dr40x9-initramfs-fit-uImage.itb

bootm

Then copy the sysupgrade image to the /tmp folder and execute sysupgrade -n <image_name>

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
Rebased and resolved the DTS conflict

robimarko commented:
@adrianschmutzler Anything blocking this from being merged?

adrianschmutzler commented:
@robimarko I normally don't merge ipqxxxx devices.

robimarko commented:
Well, it appears that nobody is.
There is quite a backlog regarding ipq40xx

adrianschmutzler commented:
There have been several devices that were merged as well, and the list of waiting devices isn't that big. It's actually relatively mediocre IMO. Just have look at lantiq for comparison, and there I closed the worst ones already ...

But putting that aside, nobody else reviewing it still won't make me capable of reviewing it myself (apart from the fact that I just won't merge each and every device support here, just because I care to give initial feedback on most).

robimarko commented:
Yes, the list is not huge but its multiple months long.
I have seen the Lantiq backlog, its been like that for quite a while now.

I understand but taking multiple months for PR-s really turns people away.

adrianschmutzler commented:
> I understand but taking multiple months for PR-s really turns people away.

So, what would you do about it?

(Don't know how it was before my time, but for me at least 1-3 months appeared to be the "normal" case.)

robimarko commented:
The only solution is more reviewers, no other way around it.

Well, in the last 3-4 years that I have been active it varied.
Sometimes it was really quick, and sometimes it took months.
It usually depends if the target area is new or really active, and sometimes it's just that maintainers are also working on the target and review happens really quickly.

adrianschmutzler commented:
> The only solution is more reviewers, no other way around it.

I tend to merge stuff on areas where I'm not so active if somebody (reliable) provides a "Reviewed-by", for example. In some cases, I even encouraged people to do that, e.g. we had some cases on kirkwood where that actually happened.

However, in most cases where no "committer" is available, it's also quite hard to find somebody else willing to invest the adequate amount of time.

Another option might be to have more automated testing and stuff like that, but this will still require substantial manual reviewing, of course.

robimarko commented:
Yeah, I understand your position.

Automation may help, but it's still a manual process.

T-X commented:
I tried the steps you described, booting into your initramfs and then flashing via sysupgarde (with the original u-boot in place, not the Wallys one), but this does not seem to work:

```
U-Boot 2012.07 [Chaos Calmer 15.05.1,unknown] (Jun 20 2019 - 16:50:44)

smem ram ptable found: ver: 1 len: 3
DRAM:  512 MiB
machid : 0x8010006
NAND:  ID = 9580f12c
Vendor = 2c
Device = f1
ONFI device found
SF: Detected MX25L1605D with page size 4 KiB, total 2 MiB
ipq_spi: page_size: 0x100, sector_size: 0x1000, size: 0x200000
130 MiB
MMC:   qca_mmc: 0
PCI0 Link Intialized
In:    serial
Out:   serial
Err:   serial
machid: 8010006
flash_type: 0
Hit any key to stop autoboot:  0
Creating 1 MTD partitions on "nand0":
0x000000000000-0x000004000000 : "mtd=0"
UBI: attaching mtd2 to ubi0
UBI: physical eraseblock size:   131072 bytes (128 KiB)
UBI: logical eraseblock size:    126976 bytes
UBI: smallest flash I/O unit:    2048
UBI: VID header offset:          2048 (aligned 2048)
UBI: data offset:                4096
UBI error: vtbl_check: volume table check failed: record 2, error 9
UBI error: ubi_init: cannot attach mtd2
UBI error: ubi_init: UBI error: cannot initialize UBI, error -22
UBI init error 22
qca_mmc: 0
```

Flashing the factory image via u-boot however seems to work:
```
(IPQ40xx) # tftpboot 0x84000000 openwrt-ipq40xx-generic-wallystech_dr40x9-squashfs-nand-factory.ubi
...
(IPQ40xx) # nand device 0
(IPQ40xx) # nand erase 0x0 0x08000000

NAND erase: device 0 whole chip
Erasing at 0x7fe0000 -- 100% complete.
OK
(IPQ40xx) # nand write ${fileaddr} 0x00000000 ${filesize}

NAND write: device 0 offset 0x0, size 0x760000
 7733248 bytes written: OK
(IPQ40xx) # reset
resetting ...
```

OpenWrt version: a recent commit on master (c2c701e058)

robimarko commented:
Well, of course, it does not work when your U-boot env is configured for a dual FW layout instead of a single one that Wallystech uses and one that OpenWrt is configured for.
I mean, what's the reason for changing the Wallystech U-boot?
It's the same code just with a different environment.

T-X commented:
@robimarko not quite sure what you mean. We don't use a "dual FW layout". We just have two different u-boot images we are trying out here, and we are manually refashing /dev/mtdblock7 / appsbl with that.

We have the QSDK u-boot the device was shipped with:

```
$ md5sum /dev/mtdblock7
d89062777b834271447506cee92bde1e  /dev/mtdblock7
```

 And then the u-boot image Wallys send us which they told us we would (supposedly) need for running OpenWrt:

```
$ md5sum /dev/mtdblock7
ccb1fe03e8008e9a5d0f5b210cc2caa3  /dev/mtdblock7
```

Does one these match the u-boot you are using?

robimarko commented:
Well, you are as the log you provided clearly shows that U-boot is trying to load from the first 64M of NAND instead of the whole.
I don't understand what your U-boot experimentation has to do with this board support in OpenWrt

ecsv commented:
> Well, you are as the log you provided clearly shows that U-boot is trying to load from the first 64M of NAND instead of the whole.

This looks more like the device was used with a MIBIB that is configured for only 64MB of NAND flash - so it is not the u-boot per se. So you have to fix the MIBIB to get it to load the ubi from the whole NAND. The NAND chip itself should be an 29F1G08ABAEA - which is a Micron 128MB (1Gb) - the number + suffix after 29F is the amount of bits.

You can also check the MIBIB partiton info by going into u-boot and then use:

    smem

Here is a MIBIB which could work: [mibib.1Gb.zip](https://github.com/openwrt/openwrt/files/5188432/mibib.1Gb.zip)

T-X reviewed (commented):
> Well, you are as the log you provided clearly shows that U-boot is trying to load from the first 64M of NAND instead of the whole.
> I don't understand what your U-boot experimentation has to do with this board support in OpenWrt

Hm, strange indeed. But that's the way we received the device. Would it be possible for you to share your MIBIB and u-boot/appsbl partitions? So that we are on the same page.

professor-jonny commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,376 @@
+
+			partition@0 {
+				label = "ubi";
+				reg = <0x00000000 0x08000000>;
```
Jason (sunbenku) the owner of wallys tech put up a fork of branch  19.07.2 with support for some of his devices as below:
https://github.com/Wallystech/openwrt

In the readme there is an upgrade procedure for converting from QSDK to vanilla Openwrt.
I have sucessfully built from the source.

dongvan9 replied:
> Jason (sunbenku) the owner of wallys tech put up a fork of branch 19.07.2 with support for some of his devices as below:
> https://github.com/Wallystech/openwrt
I've tried this firmware and unfortunately, sfp interface is not supported. As confirmed by him (Jason), sfp is only supported with the stock firmware from Wallys - Chaos Calmer 15.05.1.

robimarko replied:
This depends on the model you received.
Mine has larger NAND as I have a 512MB model but Wallys messed the U-boot completely and I can't get it to
partition as 2x256M but rather it only works as 128M partition.

professor-jonny replied:
I have the same problem as T-X above I have a 128mb device and when proforming a sysupgrade it defaults to 64mb and corrupts the UBI. I'm guessing there is no fix for this currently?

robimarko replied:
Its either changing the MIBIB or reducing the partition size in DTS

professor-jonny replied:
So you mean the QSDK magic could be patching the dts of the written image or MIBIB?
could I verify this by dumping the onboard image after an atempted sysupgrade or something?

robimarko replied:
No, its most likely that MIBIB which holds the partition layout for U-boot and QSDK has the 0x000004000000 size hardcoded and then U-boot tries to use that only

professor-jonny replied:
This is the error I get with sys upgrade with reg = <0x00000000 0x04000000 in my dts with sysupgrade   :

`
SUNBIN 1
SUNBIN command=nand device 0 && set mtdids nand0=nand0 && set mtdparts mtdparts=nand0:0x4000000@0x0(fs),${msmparts} && ubi part fs && ubi read 0x84000000 kernel && 
Creating 1 MTD partitions on "nand0":
0x000000000000-0x000004000000 : "mtd=0"
UBI: attaching mtd2 to ubi0
UBI: physical eraseblock size:   131072 bytes (128 KiB)
UBI: logical eraseblock size:    126976 bytes
UBI: smallest flash I/O unit:    2048
UBI: VID header offset:          2048 (aligned 2048)
UBI: data offset:                4096
UBI error: ubi_read_volume_table: the layout volume was not found
UBI error: ubi_init: cannot attach mtd2
UBI error: ubi_init: UBI error: cannot initialize UBI, error -22
UBI init error 22
qca_mmc: 0
(IPQ40xx) # 
`

robimarko commented:
I will rebase this onto master as my PHY driver PR was merged and it enables SFP to work.

dongvan9 commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,376 @@
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
+	model = "Wallystech DR40X9";
```
> I think the company is named "Wallys", not "Wallystech" (or "Wallys Communications (Suzhou ) Co., LTD"). wallystech.com is just the website name, I think.
> 
> Various places should be renamed from wallystech to wallys.

Agree, the same can be seen from the stock firmware provided by Wallys.

robimarko replied:
Renamed to Wallys

robimarko commented:
I have rebased this in order for SFP to work.
I am hoping to have some time today/tommorow to investigate some bugs reported here.

robimarko commented:
Updated to include proper configuration for eth0, namely so it actually registers to the PHY used.
SFP fiber support was tested and it works.

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,376 @@
+
+&pcie0 {
+	status = "okay";
+
```
No, I just tested and it does not work.
I presume it's gotta be the reset and wake GPIO-s, but that does not make sense as I pulled these from the DTS.

robimarko replied:
Ok, got the damn thing to work.
GPIOS were correct, but the polarity was not.
Qualcomm with their ignoring of the polarity in the drivers.
It works now with the updated tree.

professor-jonny commented:
I made a fork of this to play around with and I find sysupgrade images brick the router and require flashing the factory image from uboot, also that major changes require require flashing with the wallys script or it brakes the ubi partition format or something.  (error 22 in uboot) you can flash over the same image from uboot and it works but flashing a different image (larger or smalle) will fail with ubi error.

robimarko commented:
I know, it happens because using sysupgrade won't remove the QSDK specific UBI volumes, I will change the instructions to use ubiformat for the first flash.
Changing the partition size will break everything as U-boot will still resize to the size specified by MIBIB and that will break the auto resize flag as LEB/PEB number will be completely wrong.

professor-jonny commented:
would creating QSDK compatable images help?
https://github.com/openwrt/openwrt/commit/bf4630e5adb40aa393f6bd4c560baf42de0f1957

I guess at the end of the day fixinguboot would be the best idea but there is no jtag to recover from a bad flash.
I did howeer find this which seems to have more features:
https://github.com/8devices/Caraboot-ipq

Does the 40x9 support uart serial boot ?

robimarko commented:
No, QSDK compatible sysupgrade images are a really bad idea.
Using correct MIBIB so that SMEM parses gets the correct partition layout would be a decent solution.

Caraboot-ipq is 8devices fork for their Jalapeno and Habanero modules, they all use the same QCA U-boot version and simply add their board.

No, there is no UART recovery mode or something like that.

professor-jonny commented:
I'm reluctant to mess with uboot without a simple (no solder) recovery method.
jason said he has source for uboot, I have the 128/512 mb version without SFP.

robimarko commented:
Sorry, but Jason has refused to share the U-boot source multiple times despite it being GPLv2 licensed.

pepe2k commented:
> Sorry, but Jason has refused to share the U-boot source multiple times despite it being GPLv2 licensed.

And that would never change till we stop buying and supporting their broken stuff.
My OT 2c.

Cheers!

ecsv commented:
@robimarko, @pepe2k: This code https://github.com/ecsv/u-boot-wallystech-dr4029 ? I've complained a lot about the lack of sources until I convinced them that this is not secret source code from QCA - but GPL-2 code which is also freely shared by QCA. I don't know whether @T-X actually used the source code for anything.

And also keep in mind that it contains two code dumps from them - and two additional changes from me (missing header for GCC 6 + some build instructions)

T-X commented:
@ecsv: Nope, have never used that uboot source from Wallys. I stuck with the original, factory u-boot after things seemed useable after updating the u-boot environment partition table and OpenWrt DTS. I think I had tried a precompiled u-boot (which differs from the factory u-boot, which was supposed to work better with their OpenWrt patches) from Wallys at some point, but never verified if that matches the u-boot source they send you later or if these sources work at all.

pepe2k commented:
@ecsv that's better than nothing!

adschm commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq40x9-dr40x9.dts:
```diff
@@ -0,0 +1,387 @@
+		compatible = "gpio-leds";
+
+		wlan2g {
+			label = "dr4029:green:wlan2g";
```
model prefix should be removed. just use "green:wlan2g" etc.

robimarko commented:
I am gonna close this as I see that Wally sold and shipped this device in like 6 different configurations with varying RAM and NAND sizes and with different U-boot versions as well.
Its pretty much impossible to make sure that even most of them are supported.
